File: 2019/src/main/scala/Day16.scala
import scala.collection.mutable.ArrayBuffer
object Day16 {
  val input:Array[Int] = scala.io.Source.fromFile("data/data16.txt", "UTF-8").getLines.toArray.apply(0).mkString(" ").split(" ").map(_.toInt)

  def part1(): Unit = {
    var mutate:Array[Int] = Array.fill(1000)(input).flatten
    var res:Array[String] = Array.empty
    for (i <- 0 to 100) {
      println(i)
      val t = doPattern(mutate).mkString
      if (res.contains(t)) println(t)
      res :+= t
    } 
  }

  def doPattern(mutate:Array[Int]): Array[Int] = {
    var number:Array[Int] = mutate
    val base:Array[Int] = Array(0,1,0,-1)

    for (l <- 0 until 100) {
      var newNumber:Array[Int] = Array.empty
      var i = 1
      while (i <= number.length) {

        val b:ArrayBuffer[Int] = ArrayBuffer.empty

        while (number.length >= b.length) {
          base.foreach(n => b ++= ArrayBuffer.fill(i)(n))
        }
        
        newNumber :+= math.abs(number.indices.map{k => b(k+1)*number(k)}.sum) % 10

        i+=1
      }
      number = newNumber
    }
    number.take(8)
  }

  def part2(): Unit = {
    var res = Vector.fill(10000)(input.toVector).flatten
    val messageOffset = input.take(7).mkString.toInt
    res = res.slice(messageOffset, res.length)
    for (k <- 1 to 100) {
      var newRes: Vector[Int] = Vector.empty
      var sum = 0
      for (i <- res.length - 1 to 0 by -1) {
        sum += res(i)
        newRes = math.abs(sum % 10) +: newRes
      }
      res = newRes
      println(k + "/100")
    }
    println(res.take(8).mkString)
  }

  def apply() = {
    println("SOLUTION DAY 16")
    println("Running part 1")
    //part1()
    println("Running part 2")
    part2()
  }
}

File: 2019/python/src/day16.py

import os
import sys

def main():
    input = []
    
    with open(f'{os.path.dirname(os.path.realpath(__file__))}/../input/day16.txt') as f:
        for line in f.readlines():
            input.append(line.replace("\n",""))

    print("Part 1: " + str(part1(input)))
    print("Part 2: " + str(part2(input)))

def part1(input):
    signal = list([int(i) for i in input[0]])
    base = [0, 1, 0, -1]

    for _ in range(100):
        for i in range(len(signal)):
            mod_base = [b for b in base for _ in range(i+1)]
            new = 0
            for k in range(i, len(signal)):
                new += signal[k]*mod_base[(k+1)%len(mod_base)]
            signal[i] = abs(new) % 10

    return "".join([str(n) for n in signal[:8]])

def part2(input):
    signal = list(map(int, input[0])) * 10000
    offset = int(input[0][:7])

    for _ in range(100):
        num = sum([signal[i] for i in range(offset, len(signal))])
        for i in range(offset, len(signal)):
            tmp = abs(num) % 10
            num -= signal[i]
            signal[i] = tmp

    return "".join(str(n) for n in signal[offset:offset+8])

if __name__ == "__main__":
    sys.exit(main())